tests/csapi/thread_notifications_test.go

```diff
@@ -0,0 +1,178 @@
+		if !unreadNotifications.Exists() {
+			return fmt.Errorf("syncHasUnreadNotifs(%s): missing notifications", roomID)
+		}
+		if check(unreadNotifications, unreadThreadNotifications) {
+			return nil
+		}
+		return fmt.Errorf("syncHasUnreadNotifs(%s): check function did not pass: %v / %v", roomID, unreadNotifications.Raw, unreadThreadNotifications.Raw)
+	}
+}
+
+// Test behavior of threaded receipts and notifications.
+//
+// Send a series of messages, some of which are in threads. Send combinations of
+// threaded and unthreaded receipts and ensure the notification counts are updated
+// appropriately.
+func TestThreadedReceipts(t *testing.T) {
+	runtime.SkipIf(t, runtime.Dendrite) // not supported
+	deployment := Deploy(t, b.BlueprintOneToOneRoom)
+	defer deployment.Destroy(t)
+
+	// Create a room with alice and bob.
+	alice := deployment.Client(t, "hs1", "@alice:hs1")
+	bob := deployment.Client(t, "hs1", "@bob:hs1")
+
+	roomID := alice.CreateRoom(t, map[string]interface{}{"preset": "public_chat"})
+	bob.JoinRoom(t, roomID, nil)
+
+	token := bob.MustSyncUntil(t, client.SyncReq{}, client.SyncJoinedTo(bob.UserID, roomID))
+
+	// Send messages as alice and then check the highlight and notification counts from bob.
+	firstEventID := alice.SendEventSynced(t, roomID, b.Event{
+		Type: "m.room.message",
+		Content: map[string]interface{}{
+			"msgtype": "m.text",
+			"body":    "Hello world!",
+		},
+	})
+
+	// Create some threaded messages.
+	firstThreadEventID := alice.SendEventSynced(t, roomID, b.Event{
+		Type: "m.room.message",
+		Content: map[string]interface{}{
+			"msgtype": "m.text",
+			"body":    "Start thread!",
+			"m.relates_to": map[string]interface{}{
+				"event_id": firstEventID,
+				"rel_type": "m.thread",
+			},
+		},
+	})
+	alice.SendEventSynced(t, roomID, b.Event{
+		Type: "m.room.message",
+		Content: map[string]interface{}{
+			"msgtype": "m.text",
+			"body":    fmt.Sprintf("Thread response %s!", bob.UserID),
+			"m.relates_to": map[string]interface{}{
+				"event_id": firstEventID,
+				"rel_type": "m.thread",
+			},
+		},
+	})
+
+	// Send a highlight message and re-check counts.
+	secondEventID := alice.SendEventSynced(t, roomID, b.Event{
+		Type: "m.room.message",
+		Content: map[string]interface{}{
+			"msgtype": "m.text",
+			"body":    fmt.Sprintf("Hello %s!", bob.UserID),
+		},
+	})
+
+	// A filter to get thread notifications.
+	threadFilter := `{"room":{"timeline":{"unread_thread_notifications":true}}}`
+
+	// Check the unthreaded and threaded counts.
+	bob.MustSyncUntil(
+		t, client.SyncReq{Since: token},
+		client.SyncTimelineHas(roomID, func(r gjson.Result) bool {
+			return r.Get("event_id").Str == secondEventID
+		}),
+		syncHasUnreadNotifs(roomID, func(r gjson.Result, t gjson.Result) bool {
+			return r.Get("highlight_count").Num == 2 && r.Get("notification_count").Num == 4 && !t.Exists()
```

Comment:
Why is highlight count `2` when we only sent 1 highlight message according to the comments above?

Reply:
I guess from `firstEventID` and `secondEventID` but we should indicate that it's also a highlight message for `bob` in the comments.

Reply:
No, we send a highlight both in the thread and out of the thread.

I don't see anything saying that there is only a single highlight message in the comments, but I'll try to make it a bit clearer.